### PR TITLE
Fix accessing to toga layer

### PR DESCRIPTION
**Pull request by MuhammadMouradG**
## Describe your changes in detail
This PR will enable us to access the `_impl` layer from `native` layer as the documentation describe.
## What problem does this change solve?
This will fiexs #1079.

## PR Checklist:
- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
This is definitely part of 

1. There are plenty of examples where `_impl` is currently being set in the `create()` method of individual widgets. These need to be removed.

2. This would be a good opportunity to audit that all widgets do, in fact, have a `native`. They *should*... but if they don't, this patch will break.

3. There will be widgets that contain native versions *in addition* to `native` (i.e., the case of widgets that have more than one native implementation). These should also be audited and updated.

**Comment by MuhammadMouradG** [comment]
@freakboy3742 In regard to part 2 of your comment, I found in `android` implementation in `image` widget the following:
```python
if path:
    self.native = BitmapFactory.decodeFile(str(path))
elif url:
    # Android BitmapFactory nor ImageView provide a convenient async way to fetch images by URL
    self.native = None
```
then, how we can solve this, i.e. when the widget doesn't have native layer for some reason.

**Comment by MuhammadMouradG** [comment]
Is this will be a good idea if we a put `self.native._impl = self` in if condition? Like this:
```python
if not self.native:
    self.native._impl = self
else:
    RuntimeError("{} widget must be have a native layer".format(self.interface))
```

File: src/android/toga_android/widgets/base.py [hunk 1]
```diff
@@ -10,11 +10,11 @@ def __init__(self, interface):
         self.interface._impl = self
         self._container = None
         self.native = None
-        self.native._impl = self
```
Comment (MuhammadMouradG) [inline src/android/toga_android/widgets/base.py]: These changes were found after searching in the Toga project files, and then any duplicates were removed.

**Comment by MuhammadMouradG** [comment]
@freakboy3742 In regard to part 3 of your comment, could you please give me an example of these widgets?

**Comment by freakboy3742** [comment]
Take GTK's DetailedList as an example. DetailedList has a `self.native` widget - a Gtk.ScrolledWindow. However, there's also a Gtk.TreeView contained inside that scroll window. The "native" widget is the "outermost" widget, but there are 2 native widgets in the actual implementation. (There's also a data store, and an icon renderer; but those aren't really "widgets").

**Comment by MuhammadMouradG** [comment]
> Is this will be a good idea if we put `self.native._impl = self` in if condition? Like this:
> 
> ```python
> if not self.native:
>     self.native._impl = self
> else:
>     RuntimeError("{} widget does not have a native layer yet.".format(self.interface))
> ```

@freakboy3742 What do you think about this?

**Comment by MuhammadMouradG** [comment]
> @freakboy3742 In regard to part 2 of your comment, I found in `android` implementation in `image` widget the following:
> 
> ```python
> if path:
>     self.native = BitmapFactory.decodeFile(str(path))
> elif url:
>     # Android BitmapFactory nor ImageView provide a convenient async way to fetch images by URL
>     self.native = None
> ```
> 
> So, how we can solve this, i.e. when the widget doesn't have native layer for some reason.

@freakboy3742 also what we do in this situation?

**Comment by MuhammadMouradG** [comment]
> Take GTK's DetailedList as an example. DetailedList has a `self.native` widget - a Gtk.ScrolledWindow. However, there's also a Gtk.TreeView contained inside that scroll window. The "native" widget is the "outermost" widget, but there are 2 native widgets in the actual implementation. (There's also a data store, and an icon renderer; but those aren't really "widgets").

To be sure I understood you correct, specifically for this widget we need to have the following line after `self.native._impl = self`:
```python
self.treeview._impl = self
```
If the previous is correct, now we need to make it general for any number of native layers regardless of their names. My initial thought was adding the following line after `self.native._impl = self`:
```python
if self.native.get_children():
    for child in get_children():
        child._impl = self
```
So, what do you think?

**Comment by freakboy3742** [comment]
You've got the right idea with `treeview`. The general solution you've described works in principle; however, it's going to be platform specific, because not all platforms will have a `get_children()` or equivalent on their native layer.

**Comment by MuhammadMouradG** [comment]
Hmm, what do you think about solving this problem? Are we can work around this problem by define a new attribute `self.native_layers = []` that contain all native layers of the widget? Like the following
```python
if len(self.native_layers) > 1:
    for native in self.native_layers:
        native._impl = self
elif self.native:
   self.native._impl = self
else:
   RuntimeError("{} widget does not have a native layer yet.".format(self.interface))
```
The big problem that I can see with this solution is that it requires modification in every possible widget in each platform to define `self.native_layers` or at least to check this widget have only one native layer.

**Comment by MuhammadMouradG** [comment]
@freakboy3742 what do you think about this?

**Comment by freakboy3742** [comment]
Honestly, I think that approach is nothing more than kicking the problem a little further down the road. If we've missed adding `._impl` to a native object, we're just as likely to forget to add the native object to `native_layers`. Unless there's some other particularly compelling reason why we need to be able to iterate over *all* native objects (I can't think of one, and we haven't needed that capability so far), then I don't think this actually gains anything.

**Comment by MuhammadMouradG** [comment]
I think this bug will not fix if we don't suffered from it or no one has alerted us to it. Following is my trying to enlarging it and putting it on the road this will lead us to give it a little of concern. This is an editing to the previous suggestion:
```python
if self.native and self.native_layers:
    if self.native not in self.native_layers:
        self.native_layers.append(self.native)
    for native in self.native_layers:
        native._impl = self
elif self.native and self.native_layers==None:
    RuntimeError("Native layers list attribute `self.native_layers` is not define."
                 "Please, define a list and add to it the used native layers")
else:
    RuntimeError("{} widget does not have a native layer yet.".format(self.interface))
```

**Comment by freakboy3742** [comment]
Honestly - while I believe your intentions are good, I think you're massively overthinking something that actually isn't a problem that is as big as you're making it seem. 

Put it this way: we have two options:
1. A moderately complicated auto-assigning mechanism that will look for native widgets, and raise errors if they're not assigned properly. No matter how complicated we make the mechanism, there will be edge cases that it doesn't catch or can't support (native widget that isn't called native; native widgets that can't have extra attributes assigned, ...). This is code that needs to be written, debugged, and maintained over time.
2. We audit the code now, and add `_impl` to all the native widgets we find. Sometimes we'll slip up and forget to add `_impl` on a new widget; we treat that as a bug and fix it.

The thing is - this is something that is only protection against error on the part of the *Toga developers*. Checking that the Toga implementation is correct *on every single user's app execution* means everyone in the community is performing a check that we could perform once, in the factory, before we ship the next version. It's not something that is saving the *user* any effort, and it's not preventing *user* error. It won't be 100% reliable, and it will take CPU cycles to perform these automated checks. 

If anything, this is a job for static code analysis, not a runtime mechanism. However, static analysis looking for native widgets will be difficult; so I'd suggest the simple approach of "audit and fix" will be more than sufficient.

**Comment by MuhammadMouradG** [comment]
Well, to summarize what we will do:

We need to add the following line in each widget of each platform:
```python
self.native._impl = self
```
@freakboy3742 Please, confirm this if accepted, I will close this and open a new PR if this conclusion is correct.

**Comment by freakboy3742** [comment]
Yes - although, as I indicated in point (3) on my original post on this PR, it's not *just* `self.native` - some widgets have other "native" components that will also require an `_impl` attribute.

**Pull request closed without merging**